### PR TITLE
Container Initialization for caches and connections

### PR DESCRIPTION
**Pull request by simplynaveen20**
This PR will initializes the container by warming up the caches and connections.

Usage for async client - 
```
cosmosAsyncContainer.openConnectionsAndInitCaches()
cosmosContainer.openConnectionsAndInitCaches()
```

Note: This PR will open connections on first read region, if write region is different , its caches/connections wont be initialized(We we covers this as well eventually)

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:437]: I added strong consistency to cover as many connections as it can

Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:437]: how is it going to work if the account default consistency is not strong?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:435]: could you use SqlSpecQuery rather than string interpolation.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:424]: none of the APIs in this class have "Async" suffix.
please remove to be consistent.

Comment (mbhaskar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:424]: We do not add Async suffix to methods. This method is on async container already so initialize() should do.

**Review by mbhaskar — CHANGES_REQUESTED** [review]

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:443]: you just need the first page. why do you need to iterate over all pages?

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosContainerInitializationTest.java
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/CosmosContainerInitializationTest.java:5]: license header is not correct. Please check the other files.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:426]: even in GW mode many caches need to be populated which are needed for reducing query execution time later.

CollectionCache
PartitionKeyRangeCache

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (mbhaskar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:430]: Using this.getFeedRanges() instead would be cleaner?

**Review by moderakh — CHANGES_REQUESTED** [review]
1) even in GW mode many caches need to be populated which are needed for reducing query execution time later.
`CollectionCache`
`PartitionKeyRangeCache`

2) For warming up you just need first page of the query, you won't need to exhaust the query results.

3) CosmosASyncContainer methods do not have "async" suffix. please drop the `async` suffix.

4) `warmup()` seems a better method name to me. as it might be confusing what initialize mean.
when working on a structure which has an initialize method the common sense is that `initialize()` needs to be invoked. as a must where as here it is different, this is optional warmup work so it is better to rename to be more indicative of it is warmup.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:424]: warmup seems a better method name to me. as it might be confusing what initialize mean.

when working on a structure which have a initialize method the common sense is `initialize()` needs to be invoked. where as here it is different, this is optional warmup work so it is better to rename to be more indicative of it is warmup.

Comment (milismsft) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:424]: I see "warmup" as a more confusing choice then initialize... what is the .NET choice for the API name? 
Alternative is to name the API something close to what it is meant to do, which is openConnectionsToService() ;-)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (milismsft) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:439]: This particular operation consumes RUs; translation: the customer is charged $$ for something that he might not be aware of. We need appropriate documentation to capture this. Also from the implementation point of view please be sure this is the cheapest way to initialize these connections.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:424]: I am fine with `openConnectionsToService()`

Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java]: as decided in scrum today going with openConnectionsAndInitCaches

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java]: Done

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java]: I wanted to avoid non related beta apis dependency on each other. However no strong opinion on that therefore did the suggested change

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java]: Done as per offline conversation

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java]: done

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java]: It is working for direct mode(where we need this), however removed it on gateway where service is doing validation on header and throwing 400

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:430]: if the underlying container partitionKeyPath is "/id" then this query will act as a single partition query and we will only warming up the connections to a single parition.

**Review by moderakh — CHANGES_REQUESTED** [review]
`select * from c where c.id = @id`

if the underlying container partitionKeyPath is "/id" then this query will act as a single partition query and we will be only warming up the connections to a single partition.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:430]: @mbhaskar thoughts ?

Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:430]: Thats true. If "id" is partition key then `select * from c where c.id = @id` its prettymuch a single partition query

Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java]: I tested , it is not a single partition but yes also not covering all the partitions. Weird as it sound . Reverted to feed range query

**Review by moderakh — CHANGES_REQUESTED** [review]
Please try to remove the `List` and `AtomicReference` in the implementation. That adds to the complexity and makes it hard to read.

You don't really need any of that.  This can be implemented just using `flatMap` and `merge` without any `AtomicReference`, `List`, or worrying about thread-safety.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:433]: Its in lamda that's why used AtomicReference. All flux need to run in parallel for perf and I doesn't want changing.
Please let me know what is harm in this method, also give your exact code, provide perf data base comparison between two approaches. Then i will think over it , otherwise I will keep it as it is

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:432]: Redundant comment

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:431]: > Please let me know what is harm with this approach, also give your exact code, provide perf data base comparison between two approaches. Then i will think over it , otherwise I will keep it as it is

are you asking me to write the exact code and perf test the suggested approach Naveen? :-)

anyway you can take a look at cross-partition query pipeline where we are parallelizing query to multiple partitions. Over there the code has a lot of other knobs as it needs to be more configurable. Your case is simpler.

Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:431]: My previous question still stand valid "what is harm with this approach".

Are you refereeing to drainAsync method in query ? 
Based on this i see we are using same list with mergeSequential and passing our custome concurrency https://github.com/Azure/azure-sdk-for-java/blob/663620a678dc125339d7dd44d3fdbcb263ca5544/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/ParallelDocumentQueryExecutionContext.java#L388

In this case merge is better the mergeSequential , as we don't need any ordering.
And instead of using custom concurrency i would let reactor to handle 
`	/**
	 * A small default of available slots in a given container, compromise between intensive pipelines, small
	 * subscribers numbers and memory use.
	 */
	public static final int SMALL_BUFFER_SIZE = Math.max(16,
			Integer.parseInt(System.getProperty("reactor.bufferSize.small", "256")));`

Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:431]: Naveen, I added this comment when you had AtomicReference and two lists side by side here this commit:
https://github.com/Azure/azure-sdk-for-java/pull/20075/commits/2e75bb277b78f5806cc2d6167fc6cfa98915b974

my comment was more on the AtomicReference and using two lists side by side with custom synchronization logic for ensuring thread safety. I see you addressed part of my comment in a recent commit and removed the custom synchronization AtomicReference:
https://github.com/Azure/azure-sdk-for-java/pull/20075/commits/478c60c3c9398e01c746aacae13b7c416276410a

The intuition behind why I asked to remove AtomicReference, is to rely more on the reactive-stream functionality. If reactive-stream library is providing a functionality we should try to re-use and not re-invent the same concept. This helps in reducing maintenance cost and make code more readable. 

after AtomicReference is removed, there is no problem with using `fluxList` just please move its declaration inside the `flatMap`.

Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java]: Yes I already removed AtomicReference in my previous commit. Also moved fluxList  in flat map in a new Iteration. Thanks

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java]: Remove the strong consistency as per internal conversation

**Review by moderakh — APPROVED** [review]
LGTM as per conversation.

a few limitations we should be aware of.

1) If the account is multi-region the behavoiur of `openConnectionsAndInitCaches()` is not well defined with respect to which region's caches will be populated. (in case of sdk has already failed over to other regions, etc)
2) for single region account, this populates the caches which will hugely improve latency, but does not open all the connections, it only opens a connection to a single secondary.
3) the connection to primary for write will not be opened.

@FabianMeiswinkel @milismsft @kushagraThapar @simplynaveen20  please knowing the above review the documentation.

**Review by FabianMeiswinkel — CHANGES_REQUESTED** [review]
I don't think adding this at this point as new beta Api is the right apporach.

The method is only using public Api - and we know for sure the behavior/implementation will change in a way that technically speaking is breaking (increasing RU)

This is exactly the case teh Central team made. Rushing in an unstable Api just because of one unhappy customer.

So I am blocking this PR - continue with the CTL - when the results are satisfying LinkedIn can either use the code in this method (all public API) directly or we add it in implementation package and they can call into the implementation package.
But at this point I think it is not the right approach to introduce a new Beta annotated API for this when we know that RU charge etc. would change with the NorthStar implementation later.

**Review by mbhaskar — APPROVED** [review]
LGTM. Lets just make sure we document the impact of this API properly so that users can make a calculated decision. Also, publicly documenting the impact and corresponding TSGs should help. Also the flag to stop multiple inits may not be of much use

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:434]: Please add a comment on how customers should handle eventual errors - with the new logic below any retry after receiving an error on the first attempt would be a no-op.

Which means customers have two options:
1) On error crash the process - any retry would happen by trying to restart the app/service on a new process (I assume this would be the most common case - after call this method only makes sense if being called during app initialization before the health check/watch dog signal to the load balancer that it is able to accept incoming requests

2) Swallow the exception and continue (without preinitialized caches just hope lazy load will work)

Thanks

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:452]: Would it make sense to use the merge(fluxList, maxConcurrency) overload to limit concurrency? Or is it ok to just rely on the default (256)

https://github.com/reactor/reactor-core/blob/027563ecc803ad3c8afb1420f901870a10def49e/reactor-core/src/main/java/reactor/util/concurrent/Queues.java#L85-L90

**Review by FabianMeiswinkel — APPROVED** [review]
Please add the requested guidance on how to deal with transient errors that might be thrown form this method in the JavaDoc - otherwise LGTM

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:460]: Thanks for suggesting . lets go with default first see if we need any fine tune , I went over this before in this conversation https://github.com/Azure/azure-sdk-for-java/pull/20075#discussion_r601786053

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java
Comment (simplynaveen20) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java:769]: @srnagar please review this api

**Review by kushagraThapar — APPROVED** [review]
Thanks for the changes @simplynaveen20 - The API and documentation look good to me.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java
Comment (simplynaveen20) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosContainer.java:769]: Got the approval from Srikanta offline

**Pull request merged**
